### Load packages

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Load datasets

* Total population by country
* GDP (Current US$)
* % population who have access to electricity
* % electricity from Coal, Oil or Gas
* % electricity from Nuclear Power
* Fertility rate
* Smoking rate

In [3]:
# read csv file
df_raw = pd.read_csv('../data/ADS_project_dataset2.csv')

### Initial exploration

In [4]:
df_raw.head()

Key             Country Group  Year  AveTemperature  \
0             1960Bahamas             Bahamas  Rest  1960       25.167250   
1          1960Cape Verde          Cape Verde  Rest  1960       23.973167   
2  1960Congo Dem Republic  Congo Dem Republic  Rest  1960       23.767333   
3               1960Egypt               Egypt  Rest  1960       23.621167   
4              1960Gambia              Gambia  Rest  1960       27.573167   

       CO2  CO2prox1000  CO2prox2000  CO2prox3000  CO2prox4000    ...      \
0   0.4104       0.4104    2912.3927     189.5314      46.3240    ...       
1   0.0220       0.0220       0.0000    1010.4104    1030.4780    ...       
2   2.3193       2.3193      97.8581       2.6747    1435.4270    ...       
3  16.0410      35.5225    1147.6234     854.8186     218.3414    ...       
4   0.0183       0.0183       0.0000    1778.1539     363.2673    ...       

   CO2prox13000  Population         GDP  SmokingRate  FertilityRate  \
0           0.0      109528         NaN          NaN          4.495   
1           0.0      202310         NaN          NaN          6.885   
2           0.0    15248251  27434.3992          NaN          6.001   
3           0.0    26996533         NaN          NaN          6.716   
4           0.0      367928         NaN          NaN          5.573   

   ElecFossi  ElecNuclear  EduSpend  ForestArea  RoadTrans  
0        NaN          NaN       NaN         NaN        NaN  
1        NaN          NaN       NaN         NaN        NaN  
2        NaN          NaN       NaN         NaN        NaN  
3        NaN          NaN       NaN         NaN        NaN  
4        NaN          NaN       NaN         NaN        NaN  

[5 rows x 28 columns]

In [5]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9625 entries, 0 to 9624
Data columns (total 28 columns):
Key               9625 non-null object
Country           9625 non-null object
Group             9625 non-null object
Year              9625 non-null int64
AveTemperature    9275 non-null float64
CO2               9347 non-null float64
CO2prox1000       9579 non-null float64
CO2prox2000       9625 non-null float64
CO2prox3000       9625 non-null float64
CO2prox4000       9625 non-null float64
CO2prox5000       9625 non-null float64
CO2prox6000       9625 non-null float64
CO2prox7000       9625 non-null float64
CO2prox8000       9625 non-null float64
CO2prox9000       9625 non-null float64
CO2prox10000      9625 non-null float64
CO2prox11000      9625 non-null float64
CO2prox12000      9625 non-null float64
CO2prox13000      9625 non-null float64
Population        9625 non-null int64
GDP               7464 non-null float64
SmokingRate       966 non-null float64
FertilityRate     931

In [7]:
df_raw.describe(include='all')

Key  Country Group         Year  AveTemperature           CO2  \
count         9625     9625  9625  9625.000000     9275.000000   9347.000000   
unique        9625      175     2          NaN             NaN           NaN   
top     1972Gambia  Comoros  Rest          NaN             NaN           NaN   
freq             1       55  8524          NaN             NaN           NaN   
mean           NaN      NaN   NaN  1987.000000       18.933613    116.783079   
std            NaN      NaN   NaN    15.875333        8.694618    514.149162   
min            NaN      NaN   NaN  1960.000000      -19.857750     -0.080600   
25%            NaN      NaN   NaN  1973.000000       11.534875      1.234800   
50%            NaN      NaN   NaN  1987.000000       22.365750      7.919200   
75%            NaN      NaN   NaN  2001.000000       26.208792     52.117450   
max            NaN      NaN   NaN  2014.000000       30.127083  10328.733600   

         CO2prox1000   CO2prox2000   CO2prox3000   CO2prox4000      ...       \
count    9579.000000   9625.000000   9625.000000   9625.000000      ...        
unique           NaN           NaN           NaN           NaN      ...        
top              NaN           NaN           NaN           NaN      ...        
freq             NaN           NaN           NaN           NaN      ...        
mean      639.279512   1171.456248   1385.789744   1817.180871      ...        
std      1033.450195   1797.472048   1833.719431   2012.588213      ...        
min        -0.080600      0.000000      0.000000      0.000000      ...        
25%         1.947400     13.021700     93.755300    378.811800      ...        
50%        97.783000    386.353100    731.634700   1376.927500      ...        
75%      1010.952250   1639.933900   2051.540400   2556.262200      ...        
max     12487.918200  14634.930700  15088.287200  14017.505200      ...        

        CO2prox13000    Population           GDP  SmokingRate  FertilityRate  \
count    9625.000000  9.625000e+03  7.464000e+03   966.000000    9313.000000   
unique           NaN           NaN           NaN          NaN            NaN   
top              NaN           NaN           NaN          NaN            NaN   
freq             NaN           NaN           NaN          NaN            NaN   
mean       20.751072  2.840584e+07  2.585853e+05    23.727329       4.157520   
std       311.217289  1.084125e+08  9.989144e+05    10.750999       2.059739   
min         0.000000  0.000000e+00  9.873554e+01     2.200000       1.076000   
25%         0.000000  1.803064e+06  4.135951e+03    15.525000       2.198000   
50%         0.000000  5.691749e+06  1.769677e+04    23.600000       3.869000   
75%         0.000000  1.686783e+07  1.334921e+05    30.775000       6.133000   
max     10328.733600  1.364270e+09  1.617746e+07    73.400000       8.866000   

          ElecFossi  ElecNuclear     EduSpend    ForestArea     RoadTrans  
count   5568.000000  1186.000000  3225.000000  4.325000e+03  1.696000e+03  
unique          NaN          NaN          NaN           NaN           NaN  
top             NaN          NaN          NaN           NaN           NaN  
freq            NaN          NaN          NaN           NaN           NaN  
mean      59.867828    21.994675     4.415458  2.300467e+05  2.996343e+05  
std       34.098259    20.128953     1.916364  8.259457e+05  9.203500e+05  
min        0.001735     0.013356     0.704140  2.200000e+00  8.000000e+01  
25%       29.960380     4.075517     3.152830  4.800000e+03  9.291750e+03  
50%       66.285002    17.562718     4.328340  3.114000e+04  6.399150e+04  
75%       92.855628    35.976680     5.454740  1.209620e+05  2.168290e+05  
max      100.000000    87.986221    44.333980  8.151356e+06  1.338300e+07  

[11 rows x 28 columns]

#### Finding
* The raw dataset contains annual average temperature for 175 countries for the period of 1960-2014.
* Some of the predictors are sparse and have missing values, which requires to further subset and impute values.

### Clean Dataset

In [17]:
# filter data by year
START_YEAR = 1990   # some countries are missing key predictors prior to 1990
END_YEAR = 2012     # no temperature data after 2012
NUM_YEAR = END_YEAR - START_YEAR + 1  # number of years to be included in the dataset
COL_DROP = ['SmokingRate', 'EduSpend', 'RoadTrans']   # columns to drop because the majority of the data is missing

In [19]:
# set index and exclude rows before 1990 or after 2012 from the main data
df_prep = df_raw[(df_raw['Year'] >= START_YEAR) & (df_raw['Year'] <= END_YEAR)] \
    .drop(COL_DROP, axis=1) \
    .set_index(['Country', 'Year'])

In [20]:
# exclude countries that dose not have any data in any one of the columns during the period 1990-2012 as we cannot impute data 
df_country_prep = df_prep.isnull().reset_index().groupby('Country').sum() != NUM_YEAR
country_list = list(df_country_prep[df_country_prep.all(axis=1)].reset_index().Country.unique())
df = df_prep.reset_index()[df_prep.reset_index().Country.isin(country_list)].set_index(['Country', 'Year'])

country_list

['Argentina',
 'Armenia',
 'Belgium',
 'Brazil',
 'Bulgaria',
 'Canada',
 'China',
 'Czech Republic',
 'Finland',
 'France',
 'Germany',
 'Hungary',
 'India',
 'Iran',
 'Japan',
 'Lithuania',
 'Mexico',
 'Netherlands',
 'Pakistan',
 'Romania',
 'Russia',
 'Slovakia',
 'Slovenia',
 'South Africa',
 'South Korea',
 'Sweden',
 'Switzerland',
 'Ukraine',
 'United Kingdom',
 'United States of America']

### Fill missing data

In [27]:
# fill missing values
df_fillna = df.copy()
df_fillna.ElecNuclear = df_fillna.ElecNuclear.fillna(0)
df_fillna = df_fillna.fillna(method='backfill')
df_clean = df_fillna.reset_index().drop(['Key', 'Group'], axis=1)
df_clean.to_csv('../data/df_clean.csv', index=False)

### Post exploration

In [22]:
df_clean.head()

Country  Year  AveTemperature       CO2  CO2prox1000  CO2prox2000  \
0  Argentina  1990       15.155083  105.8838     105.8838     199.8423   
1    Armenia  1990        9.258083   20.0813     339.6758    1885.8749   
2    Belgium  1990       10.935333  109.7015    2742.7406     139.3653   
3     Brazil  1990       25.274917  199.8423     199.8423       0.0000   
4   Bulgaria  1990       11.597667   76.9186    1935.6349    1260.4412   

   CO2prox3000  CO2prox4000  CO2prox5000  CO2prox6000     ...      \
0       0.0000       0.0000     516.4416    4846.4523     ...       
1    3685.7329    2180.5838    1331.1099     826.9028     ...       
2     346.7531    2207.2066    8021.3267    2047.3938     ...       
3       0.0000       0.0000    5362.8939    2711.4327     ...       
4       0.0000    2772.0448    2180.5838    6854.7258     ...       

   CO2prox10000  CO2prox11000  CO2prox12000  CO2prox13000  Population  \
0      714.5774     2207.2066     3511.6937           0.0    32729739   
1        0.0000        0.0000        0.0000           0.0     3538165   
2      263.6924        0.0000        0.0000           0.0     9967379   
3      828.5306     3511.6937      149.7392           0.0   149352145   
4        0.0000        0.0000        0.0000           0.0     8718289   

            GDP  FertilityRate  ElecFossi  ElecNuclear  ForestArea  
0  42420.657180          2.989  50.208908    14.349626    347930.0  
1   3588.990350          2.544  84.993245     0.000000      3350.0  
2  36292.632790          1.620  37.805156    60.777898      6670.0  
3    321.985896          2.909   4.495537     1.003945   5467050.0  
4   1377.847468          1.820  60.743694    34.799839     33270.0  

[5 rows x 23 columns]

In [23]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 23 columns):
Country           690 non-null object
Year              690 non-null int64
AveTemperature    690 non-null float64
CO2               690 non-null float64
CO2prox1000       690 non-null float64
CO2prox2000       690 non-null float64
CO2prox3000       690 non-null float64
CO2prox4000       690 non-null float64
CO2prox5000       690 non-null float64
CO2prox6000       690 non-null float64
CO2prox7000       690 non-null float64
CO2prox8000       690 non-null float64
CO2prox9000       690 non-null float64
CO2prox10000      690 non-null float64
CO2prox11000      690 non-null float64
CO2prox12000      690 non-null float64
CO2prox13000      690 non-null float64
Population        690 non-null int64
GDP               690 non-null float64
FertilityRate     690 non-null float64
ElecFossi         690 non-null float64
ElecNuclear       690 non-null float64
ForestArea        690 non-null float64
dty

In [25]:
df_clean.describe(include='all')

Country         Year  AveTemperature          CO2  CO2prox1000  \
count           690   690.000000      690.000000   690.000000   690.000000   
unique           30          NaN             NaN          NaN          NaN   
top     South Korea          NaN             NaN          NaN          NaN   
freq             23          NaN             NaN          NaN          NaN   
mean            NaN  2001.000000       11.063343   653.160431  1792.825681   
std             NaN     6.638062        6.847785  1329.043837  1255.382571   
min             NaN  1990.000000       -5.281083     2.539200   105.883800   
25%             NaN  1995.000000        8.228646    56.665000   960.449150   
50%             NaN  2001.000000       10.105958   190.557150  1796.259700   
75%             NaN  2007.000000       13.852604   528.078550  2330.460000   
max             NaN  2012.000000       25.958750  9722.221800  9722.221800   

         CO2prox2000   CO2prox3000   CO2prox4000   CO2prox5000   CO2prox6000  \
count     690.000000    690.000000    690.000000    690.000000    690.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean     1745.179593   1332.335199   2463.923507   4776.900730   3645.743574   
std      2074.415338   1835.145169   2257.954529   4351.583099   2737.841349   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%       283.569100    346.753100    782.309300   1070.454800   1501.637025   
50%       929.508550    673.192300   2253.139600   3100.107600   2749.017750   
75%      2044.508900   1809.955800   3344.137700   7485.584600   6540.470925   
max     10301.629900  11975.498300  12196.186200  17276.702100  12350.894900   

            ...       CO2prox10000  CO2prox11000  CO2prox12000  CO2prox13000  \
count       ...         690.000000    690.000000    690.000000         690.0   
unique      ...                NaN           NaN           NaN           NaN   
top         ...                NaN           NaN           NaN           NaN   
freq        ...                NaN           NaN           NaN           NaN   
mean        ...         289.665788    299.909527    224.565107           0.0   
std         ...         492.580426   1266.236585   1243.341980           0.0   
min         ...           0.000000      0.000000      0.000000           0.0   
25%         ...           0.000000      0.000000      0.000000           0.0   
50%         ...           0.000000      0.000000      0.000000           0.0   
75%         ...         391.817900      0.000000      0.000000           0.0   
max         ...        2770.638000  11556.927300  11556.927300           0.0   

          Population           GDP  FertilityRate   ElecFossi  ElecNuclear  \
count   6.900000e+02  6.900000e+02     690.000000  690.000000   690.000000   
unique           NaN           NaN            NaN         NaN          NaN   
top              NaN           NaN            NaN         NaN          NaN   
freq             NaN           NaN            NaN         NaN          NaN   
mean    1.299571e+08  2.610925e+05       1.901425   53.235515    25.966956   
std     2.862437e+08  9.578677e+05       0.780398   26.771881    21.609769   
min     1.981629e+06  2.420839e+02       1.076000    1.220884     0.000000   
25%     8.903210e+06  4.264818e+03       1.420000   35.936571     4.868481   
50%     4.047044e+07  1.989018e+04       1.669750   60.147546    23.424987   
75%     1.027302e+08  1.031994e+05       2.010000   72.175336    40.336502   
max     1.350695e+09  5.848026e+06       6.024000   97.576563    87.986221   

          ForestArea  
count   6.900000e+02  
unique           NaN  
top              NaN  
freq             NaN  
mean    8.310122e+05  
std     1.807535e+06  
min     3.310000e+03  
25%     1.999375e+04  
50%     9